### PR TITLE
dom: add missing tags

### PR DESCRIPTION
**Pull request by krichprollsch**
according with https://wpt.live/dom/nodes/Node-cloneNode.html test, we miss HTML tags.

```
// test1: createElement
create_element_and_check("a",         "HTMLAnchorElement");
create_element_and_check("abbr",      "HTMLElement");
create_element_and_check("acronym",   "HTMLElement");
create_element_and_check("address",   "HTMLElement");
create_element_and_check("area",      "HTMLAreaElement");
create_element_and_check("article",   "HTMLElement");
create_element_and_check("aside",     "HTMLElement");
create_element_and_check("audio",     "HTMLAudioElement");
create_element_and_check("b",         "HTMLElement");
create_element_and_check("base",      "HTMLBaseElement");
create_element_and_check("bdi",       "HTMLElement");
create_element_and_check("bdo",       "HTMLElement");
create_element_and_check("bgsound",   "HTMLElement");
create_element_and_check("big",       "HTMLElement");
create_element_and_check("blockquote","HTMLElement");
create_element_and_check("body",      "HTMLBodyElement");
create_element_and_check("br",        "HTMLBRElement");
create_element_and_check("button",    "HTMLButtonElement");
create_element_and_check("canvas",    "HTMLCanvasElement");
create_element_and_check("caption",   "HTMLTableCaptionElement");
create_element_and_check("center",    "HTMLElement");
create_element_and_check("cite",      "HTMLElement");
create_element_and_check("code",      "HTMLElement");
create_element_and_check("col",       "HTMLTableColElement");
create_element_and_check("colgroup",  "HTMLTableColElement");
create_element_and_check("data",      "HTMLDataElement");
create_element_and_check("datalist",  "HTMLDataListElement");
create_element_and_check("dialog",    "HTMLDialogElement");
create_element_and_check("dd",        "HTMLElement");
create_element_and_check("del",       "HTMLModElement");
create_element_and_check("details",   "HTMLElement");
create_element_and_check("dfn",       "HTMLElement");
create_element_and_check("dir",       "HTMLDirectoryElement");
create_element_and_check("div",       "HTMLDivElement");
create_element_and_check("dl",        "HTMLDListElement");
create_element_and_check("dt",        "HTMLElement");
create_element_and_check("embed",     "HTMLEmbedElement");
create_element_and_check("fieldset",  "HTMLFieldSetElement");
create_element_and_check("figcaption","HTMLElement");
create_element_and_check("figure",    "HTMLElement");
create_element_and_check("font",      "HTMLFontElement");
create_element_and_check("footer",    "HTMLElement");
create_element_and_check("form",      "HTMLFormElement");
create_element_and_check("frame",     "HTMLFrameElement");
create_element_and_check("frameset",  "HTMLFrameSetElement");
create_element_and_check("h1",        "HTMLHeadingElement");
create_element_and_check("h2",        "HTMLHeadingElement");
create_element_and_check("h3",        "HTMLHeadingElement");
create_element_and_check("h4",        "HTMLHeadingElement");
create_element_and_check("h5",        "HTMLHeadingElement");
create_element_and_check("h6",        "HTMLHeadingElement");
create_element_and_check("head",      "HTMLHeadElement");
create_element_and_check("header",    "HTMLElement");
create_element_and_check("hgroup",    "HTMLElement");
create_element_and_check("hr",        "HTMLHRElement");
create_element_and_check("html",      "HTMLHtmlElement");
create_element_and_check("i",         "HTMLElement");
create_element_and_check("iframe",    "HTMLIFrameElement");
create_element_and_check("img",       "HTMLImageElement");
create_element_and_check("input",     "HTMLInputElement");
create_element_and_check("ins",       "HTMLModElement");
create_element_and_check("isindex",   "HTMLElement");
create_element_and_check("kbd",       "HTMLElement");
create_element_and_check("label",     "HTMLLabelElement");
create_element_and_check("legend",    "HTMLLegendElement");
create_element_and_check("li",        "HTMLLIElement");
create_element_and_check("link",      "HTMLLinkElement");
create_element_and_check("main",      "HTMLElement");
create_element_and_check("map",       "HTMLMapElement");
create_element_and_check("mark",      "HTMLElement");
create_element_and_check("marquee",   "HTMLElement");
create_element_and_check("meta",      "HTMLMetaElement");
create_element_and_check("meter",     "HTMLMeterElement");
create_element_and_check("nav",       "HTMLElement");
create_element_and_check("nobr",      "HTMLElement");
create_element_and_check("noframes",  "HTMLElement");
create_element_and_check("noscript",  "HTMLElement");
create_element_and_check("object",    "HTMLObjectElement");
create_element_and_check("ol",        "HTMLOListElement");
create_element_and_check("optgroup",  "HTMLOptGroupElement");
create_element_and_check("option",    "HTMLOptionElement");
create_element_and_check("output",    "HTMLOutputElement");
create_element_and_check("p",         "HTMLParagraphElement");
create_element_and_check("param",     "HTMLParamElement");
create_element_and_check("pre",       "HTMLPreElement");
create_element_and_check("progress",  "HTMLProgressElement");
create_element_and_check("q",         "HTMLQuoteElement");
create_element_and_check("rp",        "HTMLElement");
create_element_and_check("rt",        "HTMLElement");
create_element_and_check("ruby",      "HTMLElement");
create_element_and_check("s",         "HTMLElement");
create_element_and_check("samp",      "HTMLElement");
create_element_and_check("script",    "HTMLScriptElement");
create_element_and_check("section",   "HTMLElement");
create_element_and_check("select",    "HTMLSelectElement");
create_element_and_check("small",     "HTMLElement");
create_element_and_check("source",    "HTMLSourceElement");
create_element_and_check("spacer",    "HTMLElement");
create_element_and_check("span",      "HTMLSpanElement");
create_element_and_check("strike",    "HTMLElement");
create_element_and_check("style",     "HTMLStyleElement");
create_element_and_check("sub",       "HTMLElement");
create_element_and_check("summary",   "HTMLElement");
create_element_and_check("sup",       "HTMLElement");
create_element_and_check("table",     "HTMLTableElement");
create_element_and_check("tbody",     "HTMLTableSectionElement");
create_element_and_check("td",        "HTMLTableCellElement");
create_element_and_check("template",  "HTMLTemplateElement");
create_element_and_check("textarea",  "HTMLTextAreaElement");
create_element_and_check("th",        "HTMLTableCellElement");
create_element_and_check("time",      "HTMLTimeElement");
create_element_and_check("title",     "HTMLTitleElement");
create_element_and_check("tr",        "HTMLTableRowElement");
create_element_and_check("tt",        "HTMLElement");
create_element_and_check("track",     "HTMLTrackElement");
create_element_and_check("u",         "HTMLElement");
create_element_and_check("ul",        "HTMLUListElement");
create_element_and_check("var",       "HTMLElement");
create_element_and_check("video",     "HTMLVideoElement");
create_element_and_check("unknown",   "HTMLUnknownElement");
create_element_and_check("wbr",       "HTMLElement");
```

*todo*
* [x] fix generic HTML element tags `LoadError.WrongPrototypeTemplateProvided` error
* [x] handle tags inexisting in netsurf

### CODE REVIEW

File: src/html/elements.zig [hunk 1]
```diff
@@ -5,6 +5,7 @@ const Element = @import("../dom/element.zig").Element;
 
 // HTMLElement interfaces
 pub const Interfaces = .{
+    HTMLElement,
```
Comment (krichprollsch) [inline src/html/elements.zig:8]: trying to generate w/  `HTMLElement` here returns an error `LoadError.WrongPrototypeTemplateProvided`.
I suppose this is not the right way to do.

But in the test they wrote:
```javascript
create_element_and_check("abbr",      "HTMLElement");
create_element_and_check("acronym",   "HTMLElement");
create_element_and_check("address",   "HTMLElement");
```

Should I use another type? or create a type per tag?

See below the corresponding code in the `toInterface` func:
```zig
.abbr, .acronym, .address, .article, .aside, .b, .bdi, .bdo, .bgsound, .big, .center, .cite, .code, .dd, .details, .dfn, .dt, .figcaption, .figure, .footer, .header, .hgroup, .i, .isindex, .kbd, .main, .mark, .marquee, .nav, .nobr, .noframes, .noscript, .rp, .rt, .ruby, .s, .samp, .section, .small, .spacer, .strike, .sub, .summary, .sup, .tt, .u, .wbr, ._var => .{ .HTMLElement = @as(*parser.ElementHTML, @ptrCast(elem)) },
```

File: src/netsurf.zig [hunk 2]
```diff
@@ -68,41 +68,81 @@ inline fn stringFromData(data: []const u8) *String {
...
+    acronym = 123, // TODO c.DOM_HTML_ELEMENT_TYPE_ACRONYM,
+    bgsound = 124, // TODO c.DOM_HTML_ELEMENT_TYPE_BGSOUND,
+    big = 125, // TODO c.DOM_HTML_ELEMENT_TYPE_BIG,
+    dir = 126, // TODO c.DOM_HTML_ELEMENT_TYPE_DIR,
+    marquee = 127, // TODO c.DOM_HTML_ELEMENT_TYPE_MARQUEE,
+    nobr = 128, // TODO c.DOM_HTML_ELEMENT_TYPE_NOBR,
+    noframes = 129, // TODO c.DOM_HTML_ELEMENT_TYPE_NOFRAMES,
+    spacer = 130, // TODO c.DOM_HTML_ELEMENT_TYPE_SPACER,
+    strike = 131, // TODO c.DOM_HTML_ELEMENT_TYPE_STRIKE,
+    tt = 132, // TODO c.DOM_HTML_ELEMENT_TYPE_TT,
```
Comment (krichprollsch) [inline src/netsurf.zig]: These tags don't exist in `libdom`.
How should we handle them?

File: src/html/elements.zig (hunk 1 above)
Comment (francisbouvier) [reply src/html/elements.zig:8]: For now we have just handled tags that do not implement directly `HTMLElement` but instead an interface that inherits from it, eg. `HTMLAnchorElement`.

But those tags do implements directly, eg. `document.createElement("abbr").constructor.name === "HTMLElement"`.

So I think you are right to add `HTMLElement` in the Interfaces and to handle those tags like that in `toInterface` function.

The error is a bug in jsruntime that I have fixed with this commit https://github.com/Browsercore/jsruntime-lib/commit/3b53c93cf37b79253c3ed9e38ff9aa802e832711

File: src/html/elements.zig [hunk 3]
```diff
@@ -459,21 +495,26 @@ pub fn toInterface(comptime T: type, e: *parser.Element) T {
     const elem: *align(@alignOf(*parser.Element)) parser.Element = @alignCast(e);
     const tag = parser.elementHTMLGetTagType(@as(*parser.ElementHTML, @ptrCast(elem)));
     return switch (tag) {
+        .abbr, .acronym, .address, .article, .aside, .b, .bdi, .bdo, .bgsound, .big, .center, .cite, .code, .dd, .details, .dfn, .dt, .figcaption, .figure, .footer, .header, .hgroup, .i, .isindex, .kbd, .main, .mark, .marquee, .nav, .nobr, .noframes, .noscript, .rp, .rt, .ruby, .s, .samp, .section, .small, .spacer, .strike, .sub, .summary, .sup, .tt, .u, .wbr, ._var => .{ .HTMLElement = @as(*parser.ElementHTML, @ptrCast(elem)) },
```
Comment (francisbouvier) [inline src/html/elements.zig:498]: Maybe we could use the `else` case here for a shorter declaration.

But that can be risky too, explicit is better than implicit. WDYT?

File: src/netsurf.zig (hunk 2 above)
Comment (francisbouvier) [reply src/netsurf.zig]: We could either:
- modify libdom to handle those (I don't know about the complexity here). I think it's the clean way
- or handle that at the `nodeType` function level in zig, by for example checking if the localName match with one of those tags name and return the corresponding tag

Do we know if libdom is even able to do a `createElement` on those tags?
Does it creates an `HTMLElement` or an `HTMLUnknownElement`?

File: src/html/elements.zig (hunk 3 above)
Comment (krichprollsch) [reply src/html/elements.zig:498]: yes, it's a bit verbose but clearer to be explicit here IMO.
We don't want to return elements for unknown things.

File: src/netsurf.zig (hunk 2 above)
Comment (krichprollsch) [reply src/netsurf.zig]: > Do we know if libdom is even able to do a createElement on those tags?

 `HTMLUnknownElement` yes see https://github.com/netsurf-browser/libdom/blob/master/src/html/html_document.c#L520
 
 But we can see in https://github.com/netsurf-browser/libdom/blob/master/src/html/html_document.c#L529 many tags are just HTML element returned by netsurf.
 So I think it's easy to implement them this way. I'm trying to open a PR on netsurf. But I think we should fork in GH first to go faster.

Comment (krichprollsch) [reply src/netsurf.zig]: Relates to https://github.com/Browsercore/libdom/pull/1

**Comment by krichprollsch** [comment]
Almost correct now :tada: 

```
[36mBuilding wpt...[0m
Running WPT test suite
FAIL	tests/wpt/dom/nodes/Node-cloneNode.html
createElement(a): Pass
createElement(abbr): Pass
createElement(acronym): Pass
createElement(address): Pass
createElement(area): Pass
createElement(article): Pass
createElement(aside): Pass
createElement(audio): Pass
createElement(b): Pass
createElement(base): Pass
createElement(bdi): Pass
createElement(bdo): Pass
createElement(bgsound): Pass
createElement(big): Pass
createElement(blockquote): Pass
createElement(body): Pass
createElement(br): Pass
createElement(button): Pass
createElement(canvas): Pass
createElement(caption): Pass
createElement(center): Pass
createElement(cite): Pass
createElement(code): Pass
createElement(col): Pass
createElement(colgroup): Pass
createElement(data): Pass
createElement(datalist): Pass
createElement(dialog): Pass
createElement(dd): Pass
createElement(del): Pass
createElement(details): Pass
createElement(dfn): Pass
createElement(dir): Pass
createElement(div): Pass
createElement(dl): Pass
createElement(dt): Pass
createElement(embed): Pass
createElement(fieldset): Pass
createElement(figcaption): Pass
createElement(figure): Pass
createElement(font): Pass
createElement(footer): Pass
createElement(form): Pass
createElement(frame): Pass
createElement(frameset): Pass
createElement(h1): Pass
createElement(h2): Pass
createElement(h3): Pass
createElement(h4): Pass
createElement(h5): Pass
createElement(h6): Pass
createElement(head): Pass
createElement(header): Pass
createElement(hgroup): Pass
createElement(hr): Pass
createElement(html): Pass
createElement(i): Pass
createElement(iframe): Pass
createElement(img): Pass
createElement(input): Pass
createElement(ins): Pass
createElement(isindex): Pass
createElement(kbd): Pass
createElement(label): Pass
createElement(legend): Pass
createElement(li): Pass
createElement(link): Pass
createElement(main): Pass
createElement(map): Pass
createElement(mark): Pass
createElement(marquee): Pass
createElement(meta): Pass
createElement(meter): Pass
createElement(nav): Pass
createElement(nobr): Pass
createElement(noframes): Pass
createElement(noscript): Pass
createElement(object): Pass
createElement(ol): Pass
createElement(optgroup): Pass
createElement(option): Pass
createElement(output): Pass
createElement(p): Pass
createElement(param): Pass
createElement(pre): Pass
createElement(progress): Pass
createElement(q): Pass
createElement(rp): Pass
createElement(rt): Pass
createElement(ruby): Pass
createElement(s): Pass
createElement(samp): Pass
createElement(script): Pass
createElement(section): Pass
createElement(select): Pass
createElement(small): Pass
createElement(source): Pass
createElement(spacer): Pass
createElement(span): Pass
createElement(strike): Pass
createElement(style): Pass
createElement(sub): Pass
createElement(summary): Pass
createElement(sup): Pass
createElement(table): Pass
createElement(tbody): Pass
createElement(td): Pass
createElement(template): Pass
createElement(textarea): Pass
createElement(th): Pass
createElement(time): Pass
createElement(title): Pass
createElement(tr): Pass
createElement(tt): Pass
createElement(track): Pass
createElement(u): Pass
createElement(ul): Pass
createElement(var): Pass
createElement(video): Pass
createElement(unknown): Pass
createElement(wbr): Pass
createDocumentFragment: Fail document.createDocumentFragment is not a function
createTextNode: Fail document.createTextNode is not a function
createComment: Fail document.createComment is not a function
createElement with attributes: Fail el.setAttribute is not a function
createElementNS HTML: Fail document.createElementNS is not a function
createElementNS non-HTML: Fail document.createElementNS is not a function
createProcessingInstruction: Fail document.createProcessingInstruction is not a function
createAttribute: Fail document.createAttribute is not a function
createAttributeNS: Fail document.createAttributeNS is not a function
implementation.createDocumentType: Fail Cannot read properties of undefined (reading 'createDocumentType')
implementation.createDocument: Fail Cannot read properties of undefined (reading 'createDocument')
implementation.createHTMLDocument: Fail Cannot read properties of undefined (reading 'createHTMLDocument')
node with children: Fail Cannot read properties of undefined (reading 'length')
Node with custom prototype: Pass

1/1 tests failures
run browsercore-wpt: error: the following command exited with error code 1:
/home/pierre/prs/brwsrcr/browsercore/zig-out/bin/browsercore-wpt Node-cloneNode.html 
Build Summary: 9/11 steps succeeded; 1 failed (disable with --summary none)
wpt transitive failure
+- run browsercore-wpt failure
error: the following build command failed with exit code 1:
/home/pierre/prs/brwsrcr/browsercore/zig-cache/o/08821c369d3b13b547554a79519d2305/build /usr/local/zig-0.11.0/zig /home/pierre/prs/brwsrcr/browsercore /home/pierre/prs/brwsrcr/browsercore/zig-cache /home/pierre/.cache/zig wpt -Dengine=v8 -- Node-cloneNode.html
[33mBuild ERROR[0m
make: *** [Makefile:52: wpt] Error 1
```

**Comment by krichprollsch** [comment]
@francisbouvier PR rebased, ready for review & merge

**Pull request merged**